src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -0,0 +1,253 @@
+     * It should only be used as an optimization to reduce search space. Use {@link #dataRanges()} instead to filter data.
+     *
+     * @return merged token range
+     */
+    AbstractBounds<PartitionPosition> mergeRange()
+    {
+        return mergeRange;
+    }
+
+    /**
+     * @return indexed {@code IndexContext} if index is found; otherwise return non-indexed {@code IndexContext}.
+     */
+    public IndexContext getContext(RowFilter.Expression expression)
+    {
+        StorageAttachedIndex index = getBestIndexFor(expression);
+
+        if (index != null)
+            return index.getIndexContext();
+
+        return new IndexContext(cfs.metadata().keyspace,
+                                cfs.metadata().name,
+                                cfs.metadata().partitionKeyType,
+                                cfs.metadata().comparator,
+                                expression.column(),
+                                IndexTarget.Type.VALUES,
+                                null);
+    }
+
+    public StorageAttachedIndex getBestIndexFor(RowFilter.Expression expression)
+    {
+        return cfs.indexManager.getBestIndexFor(expression, StorageAttachedIndex.class).orElse(null);
+    }
+
+    public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionController executionController)
+    {
+        if (key == null)
+            throw new IllegalArgumentException("non-null key required");
+
+        try
+        {
+            SinglePartitionReadCommand partition = SinglePartitionReadCommand.create(cfs.metadata(),
+                                                                                     command.nowInSec(),
+                                                                                     command.columnFilter(),
+                                                                                     RowFilter.none(),
+                                                                                     DataLimits.NONE,
+                                                                                     key.partitionKey(),
+                                                                                     makeFilter(key));
+
+            return partition.queryMemtableAndDisk(cfs, executionController);
+        }
+        finally
+        {
+            queryContext.checkpoint();
+        }
+    }
+
+    /**
+     * Build a {@link RangeIterator.Builder} from the given list of expressions by applying given operation (AND).
+     * Building of such builder involves index search, results of which are persisted in the internal resources list
+     *
+     * @param expressions The expressions to build range iterator from (expressions with not results are ignored).
+     *
+     * @return range iterator builder based on given expressions and operation type.
+     */
+    public RangeIterator.Builder getIndexes(Collection<Expression> expressions)
```

Comment:
```suggestion
    public RangeIterator.Builder getIntersection(Collection<Expression> expressions)
```
There are other possibilities, I guess, but what I'm straining toward is something that is just a little more informative for the caller. It could also be something like `getIndexResults()`...open to others...

Reply:
I agree this name not meaningful. I've gone with `getIndexQueryResults`. `getIntersection` only works until we have or support in when it wouldn't be returning an intersection.